internal/controller/management/serviceaccounts/serviceaccounts.go

```diff
@@ -0,0 +1,239 @@
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	logger.Debug("done reconciling ServiceAccount")
+	return ctrl.Result{}, nil
+}
+
+// ensureControllerPermissions ensure the existence of RoleBindings that grant
+// this referenced controller ServiceAccount access to read Secrets in all
+// Project namespaces.
+func (r *reconciler) ensureControllerPermissions(ctx context.Context, sa types.NamespacedName) error {
+	roleBindingName := getRoleBindingName(sa.Name)
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"roleBinding", roleBindingName,
+	)
+	logger.Debug("ensuring necessary RoleBinding in all Project namespaces")
+
+	projectList := &kargoapi.ProjectList{}
+	if err := r.client.List(ctx, projectList); err != nil {
+		return fmt.Errorf("error listing Projects: %w", err)
+	}
+
+	// Loop through each Project to create or update the corresponding
+	// RoleBinding.
+	for _, project := range projectList.Items {
+		projectLogger := logger.WithValues("project.namespace", project.Name)
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: project.Name,
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "ClusterRole",
+				Name:     controllerReadSecretsClusterRoleName,
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind:      "ServiceAccount",
+					Name:      sa.Name,
+					Namespace: r.cfg.KargoNamespace,
+				},
+			},
+		}
+		if err := r.client.Create(ctx, roleBinding); err != nil {
+			if !kubeerr.IsAlreadyExists(err) {
+				return fmt.Errorf(
+					"error creating RoleBinding %q for ServiceAccount %q in Project namespace %q: %w",
+					roleBinding.Name, sa.Name, project.Name, err,
+				)
+			}
+			if err = r.client.Update(ctx, roleBinding); err != nil {
+				return fmt.Errorf(
+					"error updating existing RoleBinding %q in Project namespace %q: %w",
+					roleBinding.Name, project.Name, err,
+				)
+			}
+			projectLogger.Debug("updated existing RoleBinding")
+		} else {
+			projectLogger.Debug("created RoleBinding")
+		}
+	}
+	logger.Debug("necessary RoleBindings exist in all Project namespaces")
+	return nil
+}
+
+// removeControllerPermissions ensure the non-existence of RoleBindings that
+// would grant the referenced controller ServiceAccount access to read Secrets
+// in all Project namespaces.
+func (r *reconciler) removeControllerPermissions(ctx context.Context, sa types.NamespacedName) error {
+	roleBindingName := getRoleBindingName(sa.Name)
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"roleBinding", roleBindingName,
+	)
+	logger.Debug("ensuring non-existence of necessary RoleBinding in all Project namespaces")
+
+	projectList := &kargoapi.ProjectList{}
+	if err := r.client.List(ctx, projectList); err != nil {
+		return fmt.Errorf("error listing Projects: %w", err)
+	}
+
+	for _, project := range projectList.Items {
+		projectLogger := logger.WithValues("project.namespace", project.Name)
```

Comment:
Is the `project.namespace` here intentional, or should it be `project` OR `namespace`?

Reply:
It was intentional. I was trying to distinguish between `serviceAccount.namespace`  and `project.namespace`, which would be the NS wherein the RB is being created.

I suppose how we deal with this depends on the resolution of [this other comment](https://github.com/akuity/kargo/pull/2761/files#r1804533377).

👀